lldb/source/Breakpoint/StopPointSiteList.cpp

```diff
@@ -0,0 +1,235 @@
+// WatchpointResource / Watchpoint
+template <>
+bool StopPointSiteList<BreakpointSite>::StopPointSiteContainsBreakpoint(
+    typename BreakpointSite::SiteID site_id, break_id_t bp_id) {
+  std::lock_guard<std::recursive_mutex> guard(m_mutex);
+  typename collection::const_iterator pos = GetIDConstIterator(site_id);
+  if (pos != m_site_list.end())
+    return pos->second->IsBreakpointAtThisSite(bp_id);
+
+  return false;
+}
+
+template <typename StopPointSite>
+void StopPointSiteList<StopPointSite>::Dump(Stream *s) const {
+  s->Printf("%p: ", static_cast<const void *>(this));
+  // s->Indent();
+  s->Printf("StopPointSiteList with %u ConstituentSites:\n",
+            (uint32_t)m_site_list.size());
+  s->IndentMore();
+  typename collection::const_iterator pos;
+  typename collection::const_iterator end = m_site_list.end();
+  for (pos = m_site_list.begin(); pos != end; ++pos)
+    pos->second->Dump(s);
+  s->IndentLess();
+}
+
+template <typename StopPointSite>
+void StopPointSiteList<StopPointSite>::ForEach(
+    std::function<void(StopPointSite *)> const &callback) {
+  std::lock_guard<std::recursive_mutex> guard(m_mutex);
+  for (auto pair : m_site_list)
+    callback(pair.second.get());
+}
+
+template <typename StopPointSite>
+bool StopPointSiteList<StopPointSite>::FindInRange(
+    lldb::addr_t lower_bound, lldb::addr_t upper_bound,
+    StopPointSiteList &site_list) const {
+  if (lower_bound > upper_bound)
+    return false;
+
+  std::lock_guard<std::recursive_mutex> guard(m_mutex);
+  typename collection::const_iterator lower, upper, pos;
+  lower = m_site_list.lower_bound(lower_bound);
+  if (lower == m_site_list.end() || (*lower).first >= upper_bound)
+    return false;
+
+  // This is one tricky bit.  The site might overlap the bottom end of
+  // the range.  So we grab the site prior to the lower bound, and check
+  // that that + its byte size isn't in our range.
+  if (lower != m_site_list.begin()) {
+    typename collection::const_iterator prev_pos = lower;
+    prev_pos--;
+    const StopPointSiteSP &prev_site = (*prev_pos).second;
+    if (prev_site->GetLoadAddress() + prev_site->GetByteSize() > lower_bound)
+      site_list.Add(prev_site);
+  }
+
+  upper = m_site_list.upper_bound(upper_bound);
+
+  for (pos = lower; pos != upper; pos++) {
+    site_list.Add((*pos).second);
+  }
+  return true;
+}
+
+template <typename StopPointSite>
+std::vector<typename StopPointSiteList<StopPointSite>::StopPointSiteSP>
+StopPointSiteList<StopPointSite>::Sites() {
+  std::vector<StopPointSiteSP> sites;
+  std::lock_guard<std::recursive_mutex> guard(m_mutex);
+  typename collection::iterator iter = m_site_list.begin();
+  while (iter != m_site_list.end()) {
+    sites.push_back(iter->second);
+    ++iter;
+  }
+
+  return sites;
+}
+
+namespace lldb_private {
+template class StopPointSiteList<BreakpointSite>;
+template class StopPointSiteList<WatchpointResource>;
+} // namespace lldb_private
```

Comment:
Although this is perfectly fine way to allow you to define your templated methods in the `.cpp` file, as far as I can tell we don't do this anywhere else in LLDB. I found some prior art in LLVM and the fact that you have exactly two instantiations helps but given that this (1) doesn't pull in all that many dependencies and (2) probably ins't included soo much that we have to worry about compile time, I would've just implemented this in the header.